--- Modules/getpath.c.orig	2016-02-25 11:41:17.657268420 +0100
+++ Modules/getpath.c	2016-02-25 11:44:15.926441656 +0100
@@ -122,6 +122,11 @@
               EXEC_PREFIX "/lib/python" VERSION "/lib-dynload"
 #endif
 
+#undef PREFIX
+#undef PYTHONPATH
+#define PREFIX "/usr"
+#define PYTHONPATH "/usr/lib/python" VERSION ":/usr/local/lib/python" VERSION
+
 #ifndef LANDMARK
 #define LANDMARK L"os.py"
 #endif
