--- setup.py~	2015-12-07 02:39:11.000000000 +0100
+++ setup.py	2016-02-23 15:07:35.315758509 +0100
@@ -367,7 +367,8 @@
         spec = importlib.util.spec_from_file_location(ext.name, ext_filename,
                                                       loader=loader)
         try:
-            importlib._bootstrap._load(spec)
+            # importlib._bootstrap._load(spec)
+            aaaa = 1
         except ImportError as why:
             self.failed_on_import.append(ext.name)
             self.announce('*** WARNING: renaming "%s" since importing it'
