--- Lib/distutils/command/build_ext.py.orig	2016-02-23 19:52:34.825132845 +0100
+++ Lib/distutils/command/build_ext.py	2016-02-23 19:48:54.363347813 +0100
@@ -677,7 +677,10 @@
         """
         from distutils.sysconfig import get_config_var
         ext_path = ext_name.split('.')
-        ext_suffix = get_config_var('EXT_SUFFIX')
+        if 'EXT_SUFFIX' in os.environ:
+          ext_suffix = os.environ['EXT_SUFFIX']
+        else:
+          ext_suffix = get_config_var('EXT_SUFFIX')
         return os.path.join(*ext_path) + ext_suffix
 
     def get_export_symbols(self, ext):
